### PR TITLE
WEB: Sort PDEPs by their number

### LINKED ISSUE #51662: WEB: Sorting of PDEPs
xref https://github.com/pandas-dev/pandas/pull/51656#issuecomment-1445523057

The list of PDEPs is growing, and we're likely to have a PDEP-10 soon. So far, we list PDEPs at https://pandas.pydata.org/about/roadmap.html, and we started sorting the PDEP list by title in #51656. As the numbers grow to 10, sorting naively by title will make that `PDEP-2 > PDEP-10`. To keep the list sorted, we have two main options, use a regex to extract the number, convert it to int, and sort by it, or use `PDEP-001`... in PDEP PR titles so the current sorting comes naturally. Personally I don't have a preference, I guess we can make the sort a bit more complex and make our life easier. But since there is some work on PDEPs, some discussion on moving to another repo... I prefer to ask, in case there are other ideas.

### PR DESCRIPTION
**Pull request by kostyafarber**
Sort PDEPs by their number rather than the title. 

- [x] closes #51662

### CODE REVIEW

File: web/pandas_web.py [hunk 1]
```diff
@@ -307,7 +307,10 @@ def roadmap_pdeps(context):
         with open(pathlib.Path(context["target_path"]) / "pdeps.json", "w") as f:
             json.dump(pdeps, f)
 
-        for pdep in sorted(pdeps["items"], key=operator.itemgetter("title")):
+        regex = "(?<=PDEP-)\\d+"
```
Comment (datapythonista) [inline web/pandas_web.py:310]: Shouldn't the regex `r'^PDEP-(\d+)'' work? This seem unnecessarily complicated and feels like it'll return the PDEP part, not the number.

File: web/pandas_web.py [hunk 2]
```diff
@@ -307,7 +307,10 @@ def roadmap_pdeps(context):
...
+        compiled_pattern = re.compile(regex)
+        get_num = lambda pdep: int(re.findall(compiled_pattern, pdep["title"])[0])
```
Comment (datapythonista) [inline web/pandas_web.py:312]: I think `compiled_pattern.match(...)` should be more appropriate here.

**Review by datapythonista — COMMENTED** [review]
Thanks @kostyafarber. The approach seems fine, I think it's the best. I added a couple of minor comments about the implenentation.

File: web/pandas_web.py (hunk 2 above)
Comment (kostyafarber) [reply web/pandas_web.py:312]: Yep you are right. Will change.

**Comment by kostyafarber** [comment]
@datapythonista thx for review. New commit has the suggested changes.

**Comment by kostyafarber** [comment]
> @datapythonista thx for review. New commit has the suggested changes.

@datapythonista being a pest :)

**Review by datapythonista — COMMENTED** [review]
Thanks @kostyafarber, very nice. Looks great, my only concern is that if one of the sorted issues doesn't use `PDEP-\d...` as the title, I think this will fail with an error difficult to debug. I'm in my phone and can't check in detail, but let me know if I'm wrong. If I'm correct, maybe we can move the lambda to a function (can be inside the same function) and raise our own exception and error message when the pattern is not found. Thanks!

**Comment by kostyafarber** [comment]
> Thanks @kostyafarber, very nice. Looks great, my only concern is that if one of the sorted issues doesn't use `PDEP-\d...` as the title, I think this will fail with an error difficult to debug. I'm in my phone and can't check in detail, but let me know if I'm wrong. If I'm correct, maybe we can move the lambda to a function (can be inside the same function) and raise our own exception and error message when the pattern is not found. Thanks!

No worries, I think that's a sensible concern, I'll investigate and let you know.

**Comment by kostyafarber** [comment]
@datapythonista re-wrote to raise when we don't get a match with (useful) msg. Cheers

**Comment by kostyafarber** [comment]
@datapythonista friendly ping.

File: web/pandas_web.py [hunk 3]
```diff
@@ -317,7 +317,20 @@ def roadmap_pdeps(context):
         ) as f:
             json.dump(pdeps, f)
 
-        for pdep in sorted(pdeps["items"], key=operator.itemgetter("title")):
+        regex = r"^PDEP-(\d+)"
+        compiled_pattern = re.compile(regex)
```
Comment (datapythonista) [inline web/pandas_web.py:321]: I'd probably make those two lines into one. Just a personal preference, up to you, but seems unnecessarily verbose.

**Review by datapythonista — APPROVED** [review]
Thank you @kostyafarber, I added couple of final suggestions, but looks good.

File: web/pandas_web.py (hunk 3 above)
Comment (kostyafarber) [reply web/pandas_web.py:321]: Happy to change. Will ping once done. Cheers.

Comment (kostyafarber) [reply web/pandas_web.py]: @datapythonista done

**Review by datapythonista — APPROVED** [review]
Thanks @kostyafarber, looks perfect. I'll leave this open a bit to see if another core dev can have a look, bit will be merge soon if nobody does.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @kostyafarber

**Pull request merged**